### DISCUSSION TITLE
Make use of local file system

### DISCUSSION THREAD

**Discussion by tomcrane** [digirati-co-uk/iiif-manifest-editor#3] [Ideas]
Using gh-pages as backing (see #1) is all very well but you run foul of caching and can't establish a nice edit/save cycle for local work.

Could the manifest editor in the browser use https://developer.mozilla.org/en-US/docs/Web/API/FileSystem to establish a local sandboxed file system for the IIIF resources you are working with?

Then, you could "save" to GitHub when you want to share it more widely, but still be able to "serve" files locally:

```
var img = document.createElement('img');
img.src = fileEntry.toURL(); // filesystem:http://example.com/temporary/myfile.png
document.body.appendChild(img);
```

**Comment by tomcrane** [comment]
http://universalviewer.io/examples/#?manifest=XXXX

What would XXXX be here? In another tab you have "saved" your manifest somewhere

`filesystem:http://blah/mylocalmanifest.json`
Is that doable?
(either by local storage or by FileSystem API)

**Reply by stephenwf** [reply]
For temporary URLs, the UV does support data URIs.

Here is the first page of Wunder, with a custom title. It only exists in this comment.

[Manifest](http://universalviewer.io/examples/#?c=&m=&s=&cv=&manifest=data%3Aapplication%2Fjson%3Bcharset%3Dutf-8%3Bbase64%2Cew0KICAiQGNvbnRleHQiOiAiaHR0cDovL2lpaWYuaW8vYXBpL3ByZXNlbnRhdGlvbi8yL2NvbnRleHQuanNvbiIsDQogICJAaWQiOiAiaHR0cHM6Ly9paWlmLndlbGxjb21lY29sbGVjdGlvbi5vcmcvcHJlc2VudGF0aW9uL3YyL2IxODAzNTcyMyIsDQogICJAdHlwZSI6ICJzYzpNYW5pZmVzdCIsDQogICJsYWJlbCI6ICJXdW5kZXIgZGVyIFZlcmVyYnVuZyAvIFN0ZXBoZW4gY3JlYXRlZCB0aGlzIiwNCiAgIm1ldGFkYXRhIjogWw0KICAgIHsNCiAgICAgICJsYWJlbCI6ICJQdWJsaWNhdGlvbi9jcmVhdGlvbiIsDQogICAgICAidmFsdWUiOiAiTXVybmF1IDsgTcO8bmNoZW4gOiBTZWJhc3RpYW4gTHV4LCBbMTk1MV0iDQogICAgfSwNCiAgICB7DQogICAgICAibGFiZWwiOiAiUGh5c2ljYWwgZGVzY3JpcHRpb24iLA0KICAgICAgInZhbHVlIjogIjMxIHBhZ2VzIDogaWxsdXN0cmF0aW9ucyA7IDE1IGNtLiINCiAgICB9LA0KICAgIHsNCiAgICAgICJsYWJlbCI6ICJDb250cmlidXRvcnMiLA0KICAgICAgInZhbHVlIjogIkJvbGxlLCBGcml0ei4iDQogICAgfSwNCiAgICB7DQogICAgICAibGFiZWwiOiAiVHlwZS90ZWNobmlxdWUiLA0KICAgICAgInZhbHVlIjogIlBhbXBobGV0cyINCiAgICB9LA0KICAgIHsNCiAgICAgICJsYWJlbCI6ICJTdWJqZWN0cyIsDQogICAgICAidmFsdWUiOiAiR2VuZXRpY3MgLSBoaXN0b3J5Ig0KICAgIH0sDQogICAgew0KICAgICAgImxhYmVsIjogIkF0dHJpYnV0aW9uIiwNCiAgICAgICJ2YWx1ZSI6ICJXZWxsY29tZSBDb2xsZWN0aW9uIg0KICAgIH0sDQogICAgew0KICAgICAgImxhYmVsIjogIkZ1bGwgY29uZGl0aW9ucyBvZiB1c2UiLA0KICAgICAgInZhbHVlIjogIjxzcGFuPllvdSBoYXZlIHBlcm1pc3Npb24gdG8gbWFrZSBjb3BpZXMgb2YgdGhpcyB3b3JrIHVuZGVyIGEiDQogICAgfQ0KICBdLA0KICAidGh1bWJuYWlsIjogew0KICAiQGlkIjogImh0dHBzOi8vaWlpZi53ZWxsY29tZWNvbGxlY3Rpb24ub3JnL3RodW1icy9iMTgwMzU3MjNfMDAwNC5KUDIvZnVsbC83MiwxMDAvMC9kZWZhdWx0LmpwZyIsDQogICJAdHlwZSI6ICJkY3R5cGVzOkltYWdlIiwNCiAgInNlcnZpY2UiOiB7DQogICAgIkBjb250ZXh0IjogImh0dHA6Ly9paWlmLmlvL2FwaS9pbWFnZS8yL2NvbnRleHQuanNvbiIsDQogICAgIkBpZCI6ICJodHRwczovL2lpaWYud2VsbGNvbWVjb2xsZWN0aW9uLm9yZy90aHVtYnMvYjE4MDM1NzIzXzAwMDQuSlAyIiwNCiAgICAicHJvZmlsZSI6ICJodHRwOi8vaWlpZi5pby9hcGkvaW1hZ2UvMi9sZXZlbDAuanNvbiIsDQogICAgInByb3RvY29sIjogImh0dHA6Ly9paWlmLmlvL2FwaS9pbWFnZSIsDQogICAgIndpZHRoIjogNzMyLA0KICAgICJoZWlnaHQiOiAxMDI0LA0KICAgICJzaXplcyI6IFsNCiAgICAgIHsNCiAgICAgICAgIndpZHRoIjogNzIsDQogICAgICAgICJoZWlnaHQiOiAxMDANCiAgICAgIH0sDQogICAgICB7DQogICAgICAgICJ3aWR0aCI6IDE0MywNCiAgICAgICAgImhlaWdodCI6IDIwMA0KICAgICAgfSwNCiAgICAgIHsNCiAgICAgICAgIndpZHRoIjogMjg2LA0KICAgICAgICAiaGVpZ2h0IjogNDAwDQogICAgICB9LA0KICAgICAgew0KICAgICAgICAid2lkdGgiOiA3MzIsDQogICAgICAgICJoZWlnaHQiOiAxMDI0DQogICAgICB9DQogICAgXQ0KICB9DQp9LA0KICAibGljZW5zZSI6ICJodHRwOi8vY3JlYXRpdmVjb21tb25zLm9yZy9saWNlbnNlcy9ieS1uYy80LjAvIiwNCiAgImxvZ28iOiAiaHR0cHM6Ly9paWlmLndlbGxjb21lY29sbGVjdGlvbi5vcmcvbG9nb3Mvd2VsbGNvbWUtY29sbGVjdGlvbi1ibGFjay5wbmciLA0KICAicmVsYXRlZCI6IHsNCiAgICAiQGlkIjogImh0dHBzOi8vd2VsbGNvbWVjb2xsZWN0aW9uLm9yZy93b3Jrcy9rcnFwOTlyOSIsDQogICAgImZvcm1hdCI6ICJ0ZXh0L2h0bWwiLA0KICAgICJsYWJlbCI6ICJXdW5kZXIgZGVyIFZlcmVyYnVuZyAvIHZvbiBGcml0eiBCb2xsZS4iDQogIH0sDQogICJzZWVBbHNvIjogew0KICAgICJAaWQiOiAiaHR0cHM6Ly9hcGkud2VsbGNvbWVjb2xsZWN0aW9uLm9yZy9jYXRhbG9ndWUvdjIvd29ya3Mva3JxcDk5cjkiLA0KICAgICJwcm9maWxlIjogImh0dHBzOi8vYXBpLndlbGxjb21lY29sbGVjdGlvbi5vcmcvY2F0YWxvZ3VlL3YyL2NvbnRleHQuanNvbiIsDQogICAgImZvcm1hdCI6ICJhcHBsaWNhdGlvbi9qc29uIiwNCiAgICAibGFiZWwiOiAiV2VsbGNvbWUgQ29sbGVjdGlvbiBDYXRhbG9ndWUgQVBJIg0KICB9LA0KICAic2VydmljZSI6IFsNCiAgICB7DQogICAgICAiQGNvbnRleHQiOiAiaHR0cDovL2lpaWYuaW8vYXBpL3NlYXJjaC8wL2NvbnRleHQuanNvbiIsDQogICAgICAiQGlkIjogImh0dHBzOi8vaWlpZi53ZWxsY29tZWNvbGxlY3Rpb24ub3JnL3NlYXJjaC92MC9iMTgwMzU3MjMiLA0KICAgICAgInByb2ZpbGUiOiAiaHR0cDovL2lpaWYuaW8vYXBpL3NlYXJjaC8wL3NlYXJjaCIsDQogICAgICAibGFiZWwiOiAiU2VhcmNoIHdpdGhpbiB0aGlzIG1hbmlmZXN0IiwNCiAgICAgICJzZXJ2aWNlIjogew0KICAgICAgICAiQGlkIjogImh0dHBzOi8vaWlpZi53ZWxsY29tZWNvbGxlY3Rpb24ub3JnL3NlYXJjaC9hdXRvY29tcGxldGUvdjEvYjE4MDM1NzIzIiwNCiAgICAgICAgInByb2ZpbGUiOiAiaHR0cDovL2lpaWYuaW8vYXBpL3NlYXJjaC8wL2F1dG9jb21wbGV0ZSIsDQogICAgICAgICJsYWJlbCI6ICJBdXRvY29tcGxldGUgd29yZHMgaW4gdGhpcyBtYW5pZmVzdCINCiAgICAgIH0NCiAgICB9LA0KICAgIHsNCiAgICAgICJAY29udGV4dCI6ICJodHRwOi8vdW5pdmVyc2Fsdmlld2VyLmlvL2NvbnRleHQuanNvbiIsDQogICAgICAiQGlkIjogImh0dHA6Ly93ZWxsY29tZWxpYnJhcnkub3JnL3NlcnZpY2UvdHJhY2tpbmdMYWJlbHMvYjE4MDM1NzIzIiwNCiAgICAgICJwcm9maWxlIjogImh0dHA6Ly91bml2ZXJzYWx2aWV3ZXIuaW8vdHJhY2tpbmctZXh0ZW5zaW9ucy1wcm9maWxlIiwNCiAgICAgICJ0cmFja2luZ0xhYmVsIjogIkZvcm1hdDogTW9ub2dyYXBoLCBJbnN0aXR1dGlvbjogbi9hLCBJZGVudGlmaWVyOiBiMTgwMzU3MjMsIERpZ2ljb2RlOiBkaWdnZW5ldGljcywgQ29sbGVjdGlvbiBjb2RlOiBuL2EiDQogICAgfSwNCiAgICB7DQogICAgICAiQGNvbnRleHQiOiAiaHR0cDovL3dlbGxjb21lbGlicmFyeS5vcmcvbGQvaWlpZi1leHQvMC9jb250ZXh0Lmpzb24iLA0KICAgICAgIkBpZCI6ICJodHRwczovL3dlbGxjb21lbGlicmFyeS5vcmcvaWlpZi9iMTgwMzU3MjMvYWNjZXNzLWNvbnRyb2wtaGludHMtc2VydmljZSIsDQogICAgICAicHJvZmlsZSI6ICJodHRwOi8vd2VsbGNvbWVsaWJyYXJ5Lm9yZy9sZC9paWlmLWV4dC9hY2Nlc3MtY29udHJvbC1oaW50cyIsDQogICAgICAiYWNjZXNzSGludCI6ICJvcGVuIg0KICAgIH0NCiAgXSwNCiAgInNlcXVlbmNlcyI6IFsNCiAgICB7DQogICAgICAiQGlkIjogImh0dHBzOi8vaWlpZi53ZWxsY29tZWNvbGxlY3Rpb24ub3JnL3ByZXNlbnRhdGlvbi92Mi9iMTgwMzU3MjMvc2VxdWVuY2VzL3MwIiwNCiAgICAgICJAdHlwZSI6ICJzYzpTZXF1ZW5jZSIsDQogICAgICAibGFiZWwiOiAiU2VxdWVuY2UgczAiLA0KICAgICAgInJlbmRlcmluZyI6IFsNCiAgICAgICAgew0KICAgICAgICAgICJAaWQiOiAiaHR0cHM6Ly9paWlmLndlbGxjb21lY29sbGVjdGlvbi5vcmcvcGRmL2IxODAzNTcyMyIsDQogICAgICAgICAgImxhYmVsIjogIlZpZXcgYXMgUERGIiwNCiAgICAgICAgICAiZm9ybWF0IjogImFwcGxpY2F0aW9uL3BkZiINCiAgICAgICAgfSwNCiAgICAgICAgew0KICAgICAgICAgICJAaWQiOiAiaHR0cHM6Ly9hcGkud2VsbGNvbWVjb2xsZWN0aW9uLm9yZy90ZXh0L3YxL2IxODAzNTcyMyIsDQogICAgICAgICAgImxhYmVsIjogIlZpZXcgcmF3IHRleHQiLA0KICAgICAgICAgICJmb3JtYXQiOiAidGV4dC9wbGFpbiINCiAgICAgICAgfQ0KICAgICAgXSwNCiAgICAgICJ2aWV3aW5nSGludCI6ICJwYWdlZCIsDQogICAgICAiY2FudmFzZXMiOiBbDQogICAgICAgIHsNCiAgICAgICAgICAiQGlkIjogImh0dHBzOi8vaWlpZi53ZWxsY29tZWNvbGxlY3Rpb24ub3JnL3ByZXNlbnRhdGlvbi9iMTgwMzU3MjMvY2FudmFzZXMvYjE4MDM1NzIzXzAwMDEuSlAyIiwNCiAgICAgICAgICAiQHR5cGUiOiAic2M6Q2FudmFzIiwNCiAgICAgICAgICAibGFiZWwiOiAiLSIsDQogICAgICAgICAgInRodW1ibmFpbCI6IHsNCiAgIkBpZCI6ICJodHRwczovL2lpaWYud2VsbGNvbWVjb2xsZWN0aW9uLm9yZy90aHVtYnMvYjE4MDM1NzIzXzAwMDEuSlAyL2Z1bGwvNzMsMTAwLzAvZGVmYXVsdC5qcGciLA0KICAiQHR5cGUiOiAiZGN0eXBlczpJbWFnZSIsDQogICJzZXJ2aWNlIjogew0KICAgICJAY29udGV4dCI6ICJodHRwOi8vaWlpZi5pby9hcGkvaW1hZ2UvMi9jb250ZXh0Lmpzb24iLA0KICAgICJAaWQiOiAiaHR0cHM6Ly9paWlmLndlbGxjb21lY29sbGVjdGlvbi5vcmcvdGh1bWJzL2IxODAzNTcyM18wMDAxLkpQMiIsDQogICAgInByb2ZpbGUiOiAiaHR0cDovL2lpaWYuaW8vYXBpL2ltYWdlLzIvbGV2ZWwwLmpzb24iLA0KICAgICJwcm90b2NvbCI6ICJodHRwOi8vaWlpZi5pby9hcGkvaW1hZ2UiLA0KICAgICJ3aWR0aCI6IDc0MiwNCiAgICAiaGVpZ2h0IjogMTAyNCwNCiAgICAic2l6ZXMiOiBbDQogICAgICB7DQogICAgICAgICJ3aWR0aCI6IDczLA0KICAgICAgICAiaGVpZ2h0IjogMTAwDQogICAgICB9LA0KICAgICAgew0KICAgICAgICAid2lkdGgiOiAxNDUsDQogICAgICAgICJoZWlnaHQiOiAyMDANCiAgICAgIH0sDQogICAgICB7DQogICAgICAgICJ3aWR0aCI6IDI5MCwNCiAgICAgICAgImhlaWdodCI6IDQwMA0KICAgICAgfSwNCiAgICAgIHsNCiAgICAgICAgIndpZHRoIjogNzQyLA0KICAgICAgICAiaGVpZ2h0IjogMTAyNA0KICAgICAgfQ0KICAgIF0NCiAgfQ0KfSwNCiAgICAgICAgICAic2VlQWxzbyI6IHsNCiAgICAgICAgICAgICJAaWQiOiAiaHR0cHM6Ly9hcGkud2VsbGNvbWVjb2xsZWN0aW9uLm9yZy90ZXh0L2FsdG8vYjE4MDM1NzIzL2IxODAzNTcyM18wMDAxLkpQMiIsDQogICAgICAgICAgICAicHJvZmlsZSI6ICJodHRwOi8vd3d3LmxvYy5nb3Yvc3RhbmRhcmRzL2FsdG8vdjMvYWx0by54c2QiLA0KICAgICAgICAgICAgImZvcm1hdCI6ICJ0ZXh0L3htbCIsDQogICAgICAgICAgICAibGFiZWwiOiAiTUVUUy1BTFRPIFhNTCINCiAgICAgICAgICB9LA0KICAgICAgICAgICJoZWlnaHQiOiAzNTQzLA0KICAgICAgICAgICJ3aWR0aCI6IDI1NjksDQogICAgICAgICAgImltYWdlcyI6IFsNCiAgICAgICAgICAgIHsNCiAgICAgICAgICAgICAgIkBpZCI6ICJodHRwczovL2lpaWYud2VsbGNvbWVjb2xsZWN0aW9uLm9yZy9wcmVzZW50YXRpb24vYjE4MDM1NzIzL2NhbnZhc2VzL2IxODAzNTcyM18wMDAxLkpQMi9wYWludGluZy9hbm5vIiwNCiAgICAgICAgICAgICAgIkB0eXBlIjogIm9hOkFubm90YXRpb24iLA0KICAgICAgICAgICAgICAibW90aXZhdGlvbiI6ICJzYzpwYWludGluZyIsDQogICAgICAgICAgICAgICJyZXNvdXJjZSI6IHsNCiAgICAgICAgICAgICAgICAiQGlkIjogImh0dHBzOi8vaWlpZi53ZWxsY29tZWNvbGxlY3Rpb24ub3JnL2ltYWdlL2IxODAzNTcyM18wMDAxLkpQMi9mdWxsLzc0MiwxMDI0LzAvZGVmYXVsdC5qcGciLA0KICAgICAgICAgICAgICAgICJAdHlwZSI6ICJkY3R5cGVzOkltYWdlIiwNCiAgICAgICAgICAgICAgICAiZm9ybWF0IjogImltYWdlL2pwZWciLA0KICAgICAgICAgICAgICAgICJoZWlnaHQiOiAxMDI0LA0KICAgICAgICAgICAgICAgICJ3aWR0aCI6IDc0MiwNCiAgICAgICAgICAgICAgICAic2VydmljZSI6IHsNCiAgICAgICAgICAgICAgICAgICJAY29udGV4dCI6ICJodHRwOi8vaWlpZi5pby9hcGkvaW1hZ2UvMi9jb250ZXh0Lmpzb24iLA0KICAgICAgICAgICAgICAgICAgIkBpZCI6ICJodHRwczovL2lpaWYud2VsbGNvbWVjb2xsZWN0aW9uLm9yZy9pbWFnZS9iMTgwMzU3MjNfMDAwMS5KUDIiLA0KICAgICAgICAgICAgICAgICAgInByb2ZpbGUiOiAiaHR0cDovL2lpaWYuaW8vYXBpL2ltYWdlLzIvbGV2ZWwxLmpzb24iLA0KICAgICAgICAgICAgICAgICAgInByb3RvY29sIjogImh0dHA6Ly9paWlmLmlvL2FwaS9pbWFnZSIsDQogICAgICAgICAgICAgICAgICAid2lkdGgiOiAyNTY5LA0KICAgICAgICAgICAgICAgICAgImhlaWdodCI6IDM1NDMNCiAgICAgICAgICAgICAgICB9DQogICAgICAgICAgICAgIH0sDQogICAgICAgICAgICAgICJvbiI6ICJodHRwczovL2lpaWYud2VsbGNvbWVjb2xsZWN0aW9uLm9yZy9wcmVzZW50YXRpb24vYjE4MDM1NzIzL2NhbnZhc2VzL2IxODAzNTcyM18wMDAxLkpQMiINCiAgICAgICAgICAgIH0NCiAgICAgICAgICBdLA0KICAgICAgICAgICJvdGhlckNvbnRlbnQiOiBbDQogICAgICAgICAgICB7DQogICAgICAgICAgICAgICJAaWQiOiAiaHR0cHM6Ly9paWlmLndlbGxjb21lY29sbGVjdGlvbi5vcmcvYW5ub3RhdGlvbnMvdjIvYjE4MDM1NzIzL2IxODAzNTcyM18wMDAxLkpQMi9saW5lIiwNCiAgICAgICAgICAgICAgIkB0eXBlIjogInNjOkFubm90YXRpb25MaXN0IiwNCiAgICAgICAgICAgICAgImxhYmVsIjogIlRleHQgb2YgcGFnZSAgLSINCiAgICAgICAgICAgIH0NCiAgICAgICAgICBdLA0KICAgICAgICAgICJ3aXRoaW4iOiAiIg0KICAgICAgICB9DQogICAgICBdDQogICAgfQ0KICBdDQp9&xywh=-2004%2C-199%2C6575%2C3936)

**Reply by sammeltassen** [reply]
This is a good way to do previews! The [Allmaps Viewer](https://viewer.allmaps.org) supports this as well (if you paste an annotation in the text box). But in the end we created an API for these annotations (which makes much more sense in that project because the original manifest can function as a source of truth). [Some explanation here](https://observablehq.com/@bertspaan/ids-and-uris-in-allmaps?collection=@bertspaan/allmaps).

**Comment by stephenwf** [comment]
Similar to how Mock service worker works: https://mswjs.io/docs/

![](https://mermaid.ink/img/eyJjb2RlIjoic2VxdWVuY2VEaWFncmFtXG5cdEJyb3dzZXIgLT4-IFNlcnZpY2UgV29ya2VyOiAxLiByZXF1ZXN0XG4gIFNlcnZpY2UgV29ya2VyIC0tPj4gbXN3OiAyLiByZXF1ZXN0IGNsb25lXG4gIG1zdyAtLT4-IG1zdzogMy4gbWF0Y2ggYWdhaW5zdCBtb2Nrc1xuICBtc3cgLS0-PiBTZXJ2aWNlIFdvcmtlcjogNC4gTW9ja2VkIHJlc3BvbnNlXG4gIFNlcnZpY2UgV29ya2VyIC0-PiBCcm93c2VyOiA1LiByZXNwb25kV2l0aChtb2NrZWRSZXNwb25zZSlcblx0XHRcdFx0XHQiLCJtZXJtYWlkIjp7InRoZW1lIjoiZGVmYXVsdCJ9LCJ1cGRhdGVFZGl0b3IiOmZhbHNlfQ)

Where you can bypass network calls and serve locally instead (from localstorage or similar).

All of these will adhere to the same-origin policy, so you couldn't paste it directly into universal viewer or mirador websites. The viewers would have to exist on the same origin.

**Reply by tomcrane** [reply]
> The viewers would have to exist on the same origin.

I think that would be a reasonable constraint for the immediate edit/view cycle, with a push to GH and a short wait when you want to share.

**Reply by sammeltassen** [reply]
But wait doesn't that mean the viewers would be part of the manifest editor build? Or at least served as part of that? That would run against the core concept of separating editing from viewing and would limit the previewing options to those packaged with the editor...

**Reply by stephenwf** [reply]
Yes, that is a limitation of this option

**Comment by sammeltassen** [comment]
Wrote this comment first and then saw this discussion: 

- https://github.com/digirati-co-uk/iiif-manifest-editor/issues/1#issuecomment-995785459

I don't think the local file system is something separate from GitHub: I think there's a strong use case for working with locally cloned repos and loading the manifests from there with the filesystem api.

**Comment by sammeltassen** [comment]
I was thinking of other ways for people to quickly serve their manifests without using Git(Hub), which can be a threshold.

Can cloud services also serve jsons? In combination with the local file api this could make previewing or quick sharing of manifests easy, as most people have already synched folders on their device.

Standard sharing link for a file called manifest.json:

- [https://www.dropbox.com/s/fqslmvqg4izm6xr/manifest.json?dl=0](https://www.dropbox.com/s/fqslmvqg4izm6xr/manifest.json?dl=0)

Changing `www` to `dl`:

- [https://dl.dropbox.com/s/fqslmvqg4izm6xr/manifest.json?dl=0](https://dl.dropbox.com/s/fqslmvqg4izm6xr/manifest.json?dl=0)

The second link [opens in Mirador](https://projectmirador.org/embed/?manifest=https://dl.dropbox.com/s/fqslmvqg4izm6xr/manifest.json?dl=0).

(Although `Access-Control-Allow-Origin` does not seem to be specified in the CORS-headers. The link also redirects to: dl.dropboxusercontent.com but this doesn't causes trouble for Mirador.)

(There's another method [described here](https://help.dropbox.com/files-folders/share/force-download) of changing `dl=0` to `raw=1` in the initial url but this creates a HTTP redirect (unique for each version of the file) which is not supported by eg Mirador. the first method is not documented ([except for this](https://www.dropboxforum.com/t5/Dropbox-files-folders/public-links-to-raw-files/td-p/110391)) and might not be universally available.)

**Comment by tomcrane** [comment]
As well as all those good options what about this:

Manifest Editor Preview Service - **built in to the tool**

- Lambda (or Cloudflare worker, or whatever) accepts anonymous POSTs from the manifest editor (or in fact, **anyone anywhere**!) - sounds bad, but…
- this worker validates that they are valid IIIF 3 JSON documents and below a certain size (this is not a free for all hosting service - you can only host IIIF manifests!)
- This validation is as computationally low cost as possible
- It generates a google doc style obfuscated URL and returns it to the browser (in the location header) e.g., iiif-preview.digirati.io/qhLkGfWtxQ80dZF4FVBnkHE27AKXoCKo1/payload-name.json
- It modifies the id and saves in an S3 bucket (or whatever platform equivalent) - but crucially just blob storage
- The manifest is available at this URL for 1hr, or 24 hrs, or whatever - that URL is just a public (read) S3 bucket. If it’s even worth the compute of clearing it out! The manifests could just stay there forever and we clear out old ones if it becomes an issue.
- Because this can only take a JSON doc, and that must be a IIIF manifest, and it must be, say, below 100KB, this can’t be abused for hosting other stuff
- This service would likely cost peanuts (and would be good advertising for the editor, we can put an about page at the root of the URL)

Even if the editor has GitHub persistence, this solves the caching issue - you might still be committing to GH on Save, but here on Preview.
This could be an MVP bit of scaffolding to solve the most annoying problem while we build other additional solutions.
You can configure it so that repeated preview of the same manifest POSTs it with the previous id so that you are updating the stored version rather than creating a new one.

It's like code sandbox. Trivial cost compared to developing other solutions. we still need the other integrations, but this is present in the vanilla editor: anyone can use it

iiif-sandbox.io (etc).

**Reply by tomcrane** [reply]
https://github.com/digirati-co-uk/iiif-manifest-editor/pull/34
